### PR TITLE
Feature/article details screen

### PR DESCRIPTION
**Pull request by MathiasRossen**

### CODE REVIEW

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        )
+    }
+
+    fun toggleFavorite() {
+        articleState.value?.let { article ->
+            compositeDisposable.add(
+                articlesUseCase.toggleFavorite(article)
+                    .subscribeOn(ioScheduler)
+                    .subscribe { isFavorited ->
+                        println("horse $isFavorited")
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt]: Unnecessary code for debugging. Remove it.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/list/ArticlesScreenViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import dk.mathiasrossen.onboardingapp.dependency_injection.annotations.UiScheduler
+import dk.mathiasrossen.onboardingapp.use_cases.ArticlesUseCase
+import dk.mathiasrossen.onboardingapp.utils.OnboardingViewModel
+import dk.mathiasrossen.onboardingapp.utils.date.DateUtils
+import io.reactivex.rxjava3.core.Scheduler
+import io.reactivex.rxjava3.core.Single
+import javax.inject.Inject
+
+@HiltViewModel
+class ArticlesScreenViewModel @Inject constructor(
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/list/ArticlesScreenViewModel.kt]: Avoid calling `viewModel` with words like `screen`. `ViewModel` should only correspond to business logic. In this case, better rename it to `ArticlesViewModel`.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/list/ArticlesScreenViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    private val articlesUseCase: ArticlesUseCase,
+    @UiScheduler
+    private val uiScheduler: Scheduler,
+    @IoScheduler
+    private val ioScheduler: Scheduler,
+    private val dateUtils: DateUtils,
+    savedStateHandle: SavedStateHandle
+) : OnboardingViewModel() {
+    private val sourceId: String = checkNotNull(savedStateHandle[SOURCE_ID_KEY])
+    private val sourceName: String = checkNotNull(savedStateHandle[SOURCE_NAME_KEY])
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/list/ArticlesScreenViewModel.kt]: `checkNotNull` is an okay implementation. However if data is null, then application crashes and application in general is not stable.  Better add default values with the thought, that if there's no data, better to show error to the user, than crash the application.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/use_cases/ArticlesUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    fun getArticles(sourceId: String, sortBy: String, from: String?): Single<List<Article>> =
+        articleRepository.getArticlesFromSource(sourceId, sortBy, from)
+
+    fun findArticle(uuid: String): Single<Article> = articleRepository.findArticle(uuid)
+
+    fun getFavoriteArticles(): Single<List<FavoriteArticle>> =
+        favoriteArticleDao.getAll()
+
+    fun getIsArticleFavorited(article: Article): Single<Boolean> =
+        Single.fromCallable { favoriteArticleDao.find(article.url) !== null }
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/use_cases/ArticlesUseCase.kt]: Do you need to check reference equality using `!==` for nullability? I don't think that you need that. You can just use `!=`.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/use_cases/ArticlesUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+) {
+    fun getArticles(sourceId: String, sortBy: String, from: String?): Single<List<Article>> =
+        articleRepository.getArticlesFromSource(sourceId, sortBy, from)
+
+    fun findArticle(uuid: String): Single<Article> = articleRepository.findArticle(uuid)
+
+    fun getFavoriteArticles(): Single<List<FavoriteArticle>> =
+        favoriteArticleDao.getAll()
+
+    fun getIsArticleFavorited(article: Article): Single<Boolean> =
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/use_cases/ArticlesUseCase.kt]: Typo in title.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/utils/OnboardingViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package dk.mathiasrossen.onboardingapp.utils
+
+import androidx.compose.runtime.mutableStateOf
+import androidx.lifecycle.ViewModel
+import io.reactivex.rxjava3.disposables.CompositeDisposable
+
+open class OnboardingViewModel(title: String = "NewsApp"): ViewModel() {
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/utils/OnboardingViewModel.kt]: As a base `viewModel` it's better to be called as `BaseViewModel`, not `OnboardingViewModel`.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt [hunk 7]
```diff
@@ -85,20 +95,36 @@ class ArticlesScreenViewModelTest {
...
+    fun init_articleIsNotFavorited() {
+        Assert.assertFalse(viewModel.articles.values.first().value)
+    }
+
+    @Test
+    fun toggleFavorite_articleIsNowFavorited() {
+        val article = mockArticles.first()
+        given(useCase.toggleFavorite(article)).willReturn(
+            Single.just(true)
+        )
```
Comment (antsls) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt]: This fits into single line.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt [hunk 8]
```diff
@@ -19,42 +19,53 @@ import java.time.LocalDate
...
+        Article.createSample()
     )
 
     @Before
     fun setup() {
         given(dateUtils.oneDayPast()).willReturn(LocalDate.parse(date))
-        given(service.getArticlesFromSource(sourceId, language, sortBy, date)).willReturn(
-            Single.just(mockArticleResponse)
+        given(useCase.getArticles(sourceId, sortBy, date)).willReturn(
+            Single.just(mockArticles)
```
Comment (antsls) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt]: This fits into single line.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt [hunk 9]
```diff
@@ -19,42 +19,53 @@ import java.time.LocalDate
...
+        )
+        given(useCase.getFavoriteArticles()).willReturn(
+            Single.just(listOf())
         )
```
Comment (antsls) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticlesScreenViewModelTest.kt]: This fits into single line.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/navigation/Screen.kt [hunk 10]
```diff
@@ -7,7 +7,8 @@ import dk.mathiasrossen.onboardingapp.R
 sealed class Screen(val route: String, @StringRes val titleResourceId: Int, @DrawableRes val iconResourceId: Int) {
     data object Sources : Screen("sources", R.string.navigation_sources_title, R.drawable.icon_source_list) {
         const val routeMain = "main"
-        const val routeArticles = "articles/{sourceId}"
+        const val routeArticles = "articles/{sourceId}?name={sourceName}"
+        const val routeArticleDetails = "articleDetails/{articleUuid}"
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/navigation/Screen.kt]: As constants these values should be named in UPPER_SNAKE_CASE.

**Review by antsls — CHANGES_REQUESTED** [review]

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/main/MainScreen.kt [hunk 11]
```diff
@@ -21,9 +27,26 @@ import dk.mathiasrossen.onboardingapp.ui.theme.OnboardingAppTheme
 fun MainScreen(sourcesScreenViewModel: SourcesScreenViewModel) {
     val navController = rememberNavController()
     OnboardingAppTheme {
+        val appBarTitle = remember { mutableStateOf("NewsApp") }
+        val appBarImageUrl = remember { mutableStateOf<String?>(null) }
         Scaffold(
             topBar = {
-                OnboardingTopAppBar()
+                val navBackStackEntry by navController.currentBackStackEntryAsState()
+                val canNavigateBack = navController.previousBackStackEntry !== null
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/main/MainScreen.kt:36]: `!=`  is more consistent than `!==`.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleDetailsViewModelTest.kt [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    private val useCase = mock<ArticlesUseCase>()
+    private val savedStateHandle = SavedStateHandle().apply {
+        set(ArticleDetailsViewModel.ARTICLE_UUID_KEY, articleUuid)
+    }
+
+    @Before
+    fun setup() {
+        given(useCase.findArticle(articleUuid)).willReturn(
+            Single.just(mockArticle)
+        )
```
Comment (antsls) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleDetailsViewModelTest.kt:30]: This fits into single line.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleDetailsViewModelTest.kt [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        given(useCase.getIsArticleFavorited(mockArticle)).willReturn(
+            Single.just(false)
+        )
```
Comment (antsls) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleDetailsViewModelTest.kt]: This fits into single line.

File: app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleDetailsViewModelTest.kt [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        Assert.assertNotNull(viewModel.articleState.value)
+        Assert.assertEquals(mockArticle, viewModel.articleState.value)
+        Assert.assertFalse(viewModel.favoriteState.value)
+    }
+
+    @Test
+    fun toggleFavorite_articleIsNowFavorited() {
+        given(useCase.toggleFavorite(mockArticle)).willReturn(
+            Single.just(true)
+        )
```
Comment (antsls) [inline app/src/test/java/dk/mathiasrossen/onboardingapp/viewmodels/ArticleDetailsViewModelTest.kt:50]: This fits into single line.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/list/ArticlesScreenViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        refresh()
+    }
+
+    fun toggleFavorite(article: Article) {
+        compositeDisposable.add(
+            articlesUseCase.toggleFavorite(article)
+                .subscribeOn(ioScheduler)
+                .subscribe { isFavorite ->
+                    articles[article]?.value = isFavorite
+                }
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/list/ArticlesScreenViewModel.kt]: Missing `onError` implementation.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    val favoriteState = mutableStateOf(false)
+
+    init {
+        compositeDisposable.add(
+            articlesUseCase.findArticle(articleUuid)
+                .subscribeOn(ioScheduler)
+                .flatMap(
+                    { article -> articlesUseCase.getIsArticleFavorited(article) },
+                    { article, isFavorite -> article to isFavorite })
+                .subscribeOn(uiScheduler)
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt]: Should here be subscribe to UI scheduler? I think you should observe here.

**Review by antsls — CHANGES_REQUESTED** [review]

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt (hunk 16 above)
Comment (MathiasRossen) [reply app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt]: Should I use observeOn? I'm having a hard time to figure out when to change between subscribeOn and observeOn

File: app/src/main/java/dk/mathiasrossen/onboardingapp/data/article/FavoriteArticleDao.kt [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+    @Query("SELECT * FROM favoritearticle")
+    fun getAll(): Single<List<FavoriteArticle>>
+
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
+    fun insert(favoriteArticle: FavoriteArticle)
+
+    @Delete
+    fun delete(favoriteArticle: FavoriteArticle)
+
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/data/article/FavoriteArticleDao.kt:23]: Line.

File: app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+import javax.inject.Inject
+
+@HiltViewModel
+class ArticleDetailsViewModel @Inject constructor(
+    private val articlesUseCase: ArticlesUseCase,
+    @IoScheduler
+    private val ioScheduler: Scheduler,
+    savedStateHandle: SavedStateHandle
+) : BaseViewModel() {
+    private val articleUuid: String = checkNotNull(savedStateHandle[ARTICLE_UUID_KEY])
```
Comment (antsls) [inline app/src/main/java/dk/mathiasrossen/onboardingapp/ui/articles/details/ArticleDetailsViewModel.kt:20]: Add default return here.

**Review by antsls — APPROVED** [review]

**Pull request merged**